pkg/proxy/filters/ratelimit.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"net/http"
+	"regexp"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// Do we need the Retry-After value to be user configurable?
+	retryAfter      = "1"
+	cleanupInterval = 1
+)
+
+type userLimiter struct {
+	limiter  *rate.Limiter
+	lastSeen time.Time
+}
+
+type RateLimitFilter struct {
+	RequestLimit int
+	BurstLimit   int
+	excludeRegex *regexp.Regexp
+	limiters     map[string]*userLimiter
+	lock         sync.Mutex
+}
+
+func NewRateLimitFilter(requestLimit, burstLimit int, excludePattern string) (*RateLimitFilter, error) {
+	var re *regexp.Regexp
+	if excludePattern != "" {
+		var err error
+		re, err = regexp.Compile(excludePattern)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &RateLimitFilter{
+		RequestLimit: requestLimit,
+		BurstLimit:   burstLimit,
+		excludeRegex: re,
```

Comment:
Yeah I did consider that, but I decided it was probably best to avoid doing any regex comparison at all unless the regex flag is specified, hence in `rateLimitUser` we just check if this is nil.

My thinking here is we might perhaps have a separate user-facing proxy endpoint (e.g one that's only configured for OIDC auth perhaps), and in that case we won't need the regex at all to allow admin/SA access through that endpoint (but we will need it for now in the case where there's one proxy instance that does handle syncer traffic etc)

Reply:
Sounds good!